### DISCUSSION TITLE
Native Hook no enabled!

### DISCUSSION THREAD

**Discussion by mohaomer12** [nfcgate/nfcgate#129] [Q&A] [answered]
Hi,
I have installed NFCgate Ver 2.2.0 on a rooted Redmi Note 9 Pro.
NXP PN553
Android Ver 11.
Build Number: RKQ1.200826.002
Riru Ver 22.4
Riru - EdXposed Ver 0.5.0.6
Status Screen showed all are ok except Native Hook Enabled = No.
Any Idea what could be the problem?

**Comment by roussosalex** [comment]
Did you reboot the device? Is the hook enabled in edXposed for the nfc service?
If the issue persists, please attach a boot log (logcat since boot) so we can diagnose the issue.

**Reply by mohaomer12** [reply]
Hi,
Thank you for quick reply,
Yes, I rebooted the device many times.
When I started the nfcgate app it is showing "Warning: Xposed module is not working properly" in the home screen.
In status screen all are ok except the Native Hook Enabled = No, and when I tap it it is showing "Warning: Native hook could not be found".
BTW, the Capture Mode is working well, but relay in tag mode doesn't work, reader start to connect to the emulated tag and after 1 second  disconnect. 
The nfc service in EdXposed Manager was not enabled, so I enabled it in EdXposed Manager-> Compat List -> com.android.nfc
I will show you below the log file of EdXposed Manager and the Logcat since the phone boot:
**1-EdXposed Manager log:**
--------- beginning of head
EdXposed Log
Powered by Log Catcher
QQ support group: 855219808
Telegram support group: @Code_Of_MeowCat
Telegram channel: @EdXposed
--------- beginning of information
Manufacturer: Xiaomi
Brand: Redmi
Device: joyeuse
Product: joyeuse
Model: Redmi Note 9 Pro
Fingerprint: Redmi/joyeuse_global/joyeuse:11/RKQ1.200826.002/V12.5.7.0.RJZMIXM:user/release-keys
ROM description: joyeuse_global-user 11 RKQ1.200826.002 V12.5.7.0.RJZMIXM release-keys
Architecture: arm64-v8a
Android build: RKQ1.200826.002
Android version: 11
Android sdk: 30
EdXposed version: v0.5.0.6_4569-android_r (YAHFA)
EdXposed api: 91.0
Riru version:  ()
Riru api: 
Magisk: 25.2 (25200)
--------- beginning of main
--------- beginning of crash
--------- beginning of system
08-30 10:05:11.979   919   919 I EdXposed: onModuleLoaded: welcome to EdXposed!
08-30 10:05:11.979   919   919 I EdXposed: Start to install inline hooks
08-30 10:05:11.979   919   919 I EdXposed: Using api level 30
08-30 10:05:11.979   919   919 I EdXposed: Start to install Riru hook
08-30 10:05:11.984   918   918 I EdXposed: onModuleLoaded: welcome to EdXposed!
08-30 10:05:11.984   918   918 I EdXposed: Start to install inline hooks
08-30 10:05:11.984   918   918 I EdXposed: Using api level 30
08-30 10:05:11.984   918   918 I EdXposed: Start to install Riru hook
08-30 10:05:12.010   919   919 I EdXposed: Riru hooks installed
08-30 10:05:12.017   918   918 I EdXposed: Riru hooks installed
08-30 10:05:12.061   919   919 I EdXposed: using installer org.meowcat.edxposed.manager
08-30 10:05:12.062   919   919 I EdXposed: data path prefix: /data/user_de/0/
08-30 10:05:12.062   919   919 I EdXposed:   application list mode: false
08-30 10:05:12.062   919   919 I EdXposed:     using whitelist: false
08-30 10:05:12.062   919   919 I EdXposed:   dynamic modules mode: false
08-30 10:05:12.062   919   919 I EdXposed:   resources hook: false
08-30 10:05:12.062   919   919 I EdXposed:   deopt boot image: false
08-30 10:05:12.062   919   919 I EdXposed:   no module log: false
08-30 10:05:12.062   919   919 I EdXposed:   hidden api bypass: true
08-30 10:05:12.064   919   919 I EdXposed: ART hooks installed
08-30 10:05:12.090   918   918 I EdXposed: using installer org.meowcat.edxposed.manager
08-30 10:05:12.090   918   918 I EdXposed: data path prefix: /data/user_de/0/
08-30 10:05:12.090   918   918 I EdXposed:   application list mode: false
08-30 10:05:12.090   918   918 I EdXposed:     using whitelist: false
08-30 10:05:12.090   918   918 I EdXposed:   dynamic modules mode: false
08-30 10:05:12.090   918   918 I EdXposed:   resources hook: false
08-30 10:05:12.090   918   918 I EdXposed:   deopt boot image: false
08-30 10:05:12.090   918   918 I EdXposed:   no module log: false
08-30 10:05:12.090   918   918 I EdXposed:   hidden api bypass: true
08-30 10:05:12.091   918   918 I EdXposed: ART hooks installed
08-30 10:05:17.248   918   918 I EdXposed-Bridge: Loading modules from /data/app/~~fc9aLoeUPxog3SHGNVqNpg==/org.meowcat.edxposed.manager-kDjbZ4878kk_sfyWmytjCw==/base.apk
08-30 10:05:17.444   918   918 I EdXposed-Bridge:   Loading class org.meowcat.edxposed.manager.xposed.Enhancement
08-30 10:05:17.445   918   918 I EdXposed-Bridge: Loading modules from /data/app/~~vkW_DMzsX4R5SVo3pEHXQw==/de.tu_darmstadt.seemoo.nfcgate-v0nIXcChrK6pWSagsgJ1Zw==/base.apk
08-30 10:05:17.553   918   918 I EdXposed-Bridge:   Loading class de.tu_darmstadt.seemoo.nfcgate.xposed.Hooks
08-30 10:05:24.947   919   919 W EdXposed: skip injecting into WebViewLoader-armeabi-v7a because it's isolated
08-30 10:05:24.947   919   919 W EdXposed: skip injecting xposed into WebViewLoader-armeabi-v7a because it's whitelisted/blacklisted
08-30 10:05:25.012   918   918 W EdXposed: skip injecting into WebViewLoader-arm64-v8a because it's isolated
08-30 10:05:25.012   918   918 W EdXposed: skip injecting xposed into WebViewLoader-arm64-v8a because it's whitelisted/blacklisted
08-30 10:05:25.100   919   919 I EdXposed-Bridge: Loading modules from /data/app/~~fc9aLoeUPxog3SHGNVqNpg==/org.meowcat.edxposed.manager-kDjbZ4878kk_sfyWmytjCw==/base.apk
08-30 10:05:25.320   919   919 I EdXposed-Bridge:   Loading class org.meowcat.edxposed.manager.xposed.Enhancement
08-30 10:05:25.321   919   919 I EdXposed-Bridge: Loading modules from /data/app/~~vkW_DMzsX4R5SVo3pEHXQw==/de.tu_darmstadt.seemoo.nfcgate-v0nIXcChrK6pWSagsgJ1Zw==/base.apk
08-30 10:05:25.404   919   919 I EdXposed-Bridge:   Loading class de.tu_darmstadt.seemoo.nfcgate.xposed.Hooks
08-30 10:05:25.437   919   919 W EdXposed: skip injecting into webview_zygote because it's a child zygote
08-30 10:05:25.437   919   919 W EdXposed: skip injecting xposed into webview_zygote because it's whitelisted/blacklisted
--------- beginning of system
--------- beginning of system
--------- beginning of system
--------- beginning of system
--------- beginning of main
--------- beginning of system
--------- beginning of main
08-30 10:06:22.582   919   919 W EdXposed: skip injecting into com.android.chrome_zygote because it's a child zygote
08-30 10:06:22.582   919   919 W EdXposed: skip injecting xposed into com.android.chrome_zygote because it's whitelisted/blacklisted
08-30 10:07:25.263   919   919 W EdXposed: skip injecting into com.android.chrome_zygote because it's a child zygote
08-30 10:07:25.263   919   919 W EdXposed: skip injecting xposed into com.android.chrome_zygote because it's whitelisted/blacklisted
08-30 10:08:20.383 13019 13019 D EdXposedManager: EdXposed is not active
08-30 10:08:20.384 13019 13019 D EdXposedManager: EdXposed is not active
08-30 10:08:20.384 13019 13019 D EdXposedManager: ApplicationList: Force add modules to list
08-30 10:08:20.776 13019 13019 D EdXposedManager: EdXposed is not active
08-30 10:09:14.236   918   918 E EdXposed: can't parse /data/system
08-30 10:09:14.236   918   918 W EdXposed: skip injecting xposed into android:ui because it's whitelisted/blacklisted
08-30 10:12:38.364   919   919 W EdXposed: skip injecting into com.android.chrome_zygote because it's a child zygote
08-30 10:12:38.364   919   919 W EdXposed: skip injecting xposed into com.android.chrome_zygote because it's whitelisted/blacklisted
08-30 10:34:37.146   918   918 E EdXposed: can't parse /data/system
08-30 10:34:37.146   918   918 W EdXposed: skip injecting xposed into android:ui because it's whitelisted/blacklisted



**USB Reader WireSahrk:**
[USB-Reader-WireSahrk.txt](https://github.com/nfcgate/nfcgate/files/12478464/USB-Reader-WireSahrk.txt)

**Pyhton-Server-Log:**
[Pyhton-Server-Log.txt](https://github.com/nfcgate/nfcgate/files/12478405/Pyhton-Server-Log.txt)

**Reply by kleest** [reply]
I do not see any "HOOKNFC" or "NATIVENFC" entries in your log so it looks like the Xposed module is not properly injected into the com.android.nfc service. After enabling the Xposed module for com.android.nfc service, you have to restart the service either by killing it (kill -9 in a root shell) or restarting the device. Turning NFC off/on is not always sufficient.

**Reply by mohaomer12** [reply]
But, the On-Device Capture mode is working well, does that mean the hook nfc is working well?
My problem now with Relay in tag mode.

**Reply by mohaomer12** [reply]
Hi 
I found some log might help:

2023-09-03 19:56:15.542  3964-3964  NATIVENFC               pid-3964                             I  Library found at /system_ext/lib64/libsn100nfc-nci.so
2023-09-03 19:56:15.545  3964-3964  NATIVENFC               pid-3964                             D  hookNFA_CB: using offset 0x40
2023-09-03 19:56:15.565  3964-3964  NATIVENFC               pid-3964                             D  hookNFA_CB: found p_nfa_conn_cback 0x0 with permissions 0
2023-09-03 19:56:15.565  3964-3964  NATIVENFC               pid-3964                             E  Error at ../../../../src/main/jni/src/nfcd.cpp:208
2023-09-03 19:56:15.565  3964-3964  NATIVENFC               pid-3964                             E  NFA_CB invalid or not read+execute
2023-09-03 19:56:15.565  3964-3964  NATIVENFC               pid-3964                             E  Error at ../../../../src/main/jni/src/nfcd.cpp:261
2023-09-03 19:56:15.565  3964-3964  NATIVENFC               pid-3964                             E  Hooking nfa_cb failed

**Reply by kleest** [reply, marked as answer]
Thank you for providing the log, it really did help us improve compatibility. We created a new [pre-release version](https://github.com/nfcgate/nfcgate/releases/tag/v2.3.0-rc). Could you give this version a try and report back if it fixes your issues? Many thanks :pray:

**Reply by mohaomer12** [reply]
Thank you🙏
Now working well!

![NFCGateTest](https://github.com/nfcgate/nfcgate/assets/4160158/3ac649d3-b5a1-4056-b698-d4b3a2970f41)

**Comment by Serhii-Tymoshchenko** [comment]


**Reply by kleest** [reply]
Please attach a boot log (logcat since boot) so we can diagnose the issue.

**Comment by ROinfo86** [comment]
I have the same problem -_-
Please HELP!
xiaomi redmi 10 2022 (V14.0.8.0.TKUEUXM)

![Screenshot_2024-06-04-14-43-49-442_com topjohnwu magisk](https://github.com/nfcgate/nfcgate/assets/38342536/711c7313-2a95-4cb0-b631-7d6d5c5c90b5)
![Screenshot_2024-06-04-14-44-02-180_com topjohnwu magisk](https://github.com/nfcgate/nfcgate/assets/38342536/e26f63da-179c-4f67-b63d-dc4ab256b23d)
![Screenshot_2024-06-04-14-44-29-282_org lsposed manager](https://github.com/nfcgate/nfcgate/assets/38342536/74d105d2-f359-4167-95ab-a3ea6844e800)
![Screenshot_2024-06-04-14-44-47-639_org lsposed manager](https://github.com/nfcgate/nfcgate/assets/38342536/0097a288-40ff-4f95-9b69-6bcf51ac1282)
![Screenshot_2024-06-04-14-46-06-082_de tu_darmstadt seemoo nfcgate](https://github.com/nfcgate/nfcgate/assets/38342536/746db12d-764d-4814-9dd1-89cc60d5846d)

-------
Logs
[magisk_log_2024-06-04T14.43.15.log](https://github.com/user-attachments/files/15550903/magisk_log_2024-06-04T14.43.15.log)
[LSPosed_2024-06-04T14_45_33.895602.zip](https://github.com/user-attachments/files/15550923/LSPosed_2024-06-04T14_45_33.895602.zip)

**Reply by kleest** [reply]
@ROinfo86 please try the newest version 2.4.0.

**Reply by ROinfo86** [reply]
Didn't help  -_-

**Reply by kleest** [reply]
Then please attach logs running with the newest version.

**Comment by motherloader** [comment]
![Screenshot_2024-06-12-03-50-34-148_de tu_darmstadt seemoo nfcgate 1](https://github.com/nfcgate/nfcgate/assets/172453233/886a3c6d-7b15-43e3-9530-0c61480cc4a7)
Redmi 12C MIUI Global 14.0.2.0(TCVRUXM) Android 13. NfcGate 2.4.2 LSPosed 1.9.2
Native Hook Enabled - orange
[magisk_log_2024-06-12T03.33.02[1].log](https://github.com/user-attachments/files/15796239/magisk_log_2024-06-12T03.33.02.1.log)
[LSPosed_2024-06-12T03_32_05.252917[1].zip](https://github.com/user-attachments/files/15796244/LSPosed_2024-06-12T03_32_05.252917.1.zip)

**Reply by kleest** [reply]
Unfortunately the logcat full.log in your log archive is incomplete and missing important information. For example, there is no "NATIVENFC" log entry.
It also looks like your phone somehow uses an old version of the library, because the latest tombstone (tombstone_25) reports a crash of the NFC service with NFCGate version 2.4.0.

Please ensure that your phone uses the correct version by manually restarting the NFC service after installing the newest NFCGate version (with `killall -9 com.android.nfc`) and attach a complete logcat that contains relevant log entries (tags "NATIVENFC" and "HOOKNFC") if the problem still occurs.
